File: core/src/ten_runtime/binding/nodejs/interface/msg/msg.ts
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//
import { TenError } from "../error/error.js";
import ten_addon from "../ten_addon.js";

export class Msg {
  getName(): string {
    return ten_addon.ten_nodejs_msg_get_name(this);
  }

  setDest(
    appUri: string | undefined = undefined,
    graphId: string | undefined = undefined,
    extension: string | undefined = undefined,
  ) {
    ten_addon.ten_nodejs_msg_set_dest(this, appUri, graphId, extension);
  }

  setPropertyFromJson(path: string, jsonStr: string): TenError | null {
    return ten_addon.ten_nodejs_msg_set_property_from_json(this, path, jsonStr);
  }

  getPropertyToJson(path: string): [string, TenError | null] {
    return ten_addon.ten_nodejs_msg_get_property_to_json(this, path);
  }

  setPropertyNumber(path: string, value: number): TenError | null {
    return ten_addon.ten_nodejs_msg_set_property_number(this, path, value);
  }

  getPropertyNumber(path: string): [number, TenError | null] {
    return ten_addon.ten_nodejs_msg_get_property_number(this, path);
  }

  setPropertyString(path: string, value: string): TenError | null {
    return ten_addon.ten_nodejs_msg_set_property_string(this, path, value);
  }

  getPropertyString(path: string): [string, TenError | null] {
    return ten_addon.ten_nodejs_msg_get_property_string(this, path);
  }

  setPropertyBool(path: string, value: boolean): TenError | null {
    return ten_addon.ten_nodejs_msg_set_property_bool(this, path, value);
  }

  getPropertyBool(path: string): [boolean, TenError | null] {
    return ten_addon.ten_nodejs_msg_get_property_bool(this, path);
  }

  setPropertyBuf(path: string, value: ArrayBuffer): TenError | null {
    return ten_addon.ten_nodejs_msg_set_property_buf(this, path, value);
  }

  getPropertyBuf(path: string): [ArrayBuffer, TenError | null] {
    return ten_addon.ten_nodejs_msg_get_property_buf(this, path);
  }
}

File: core/src/ten_manager/src/designer/exec/msg.rs
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//
use serde::{Deserialize, Serialize};

use crate::log::LogLineInfo;

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type")]
pub enum InboundMsg {
    #[serde(rename = "exec_cmd")]
    ExecCmd { base_dir: String, cmd: String, stdout_is_log: bool, stderr_is_log: bool },

    #[serde(rename = "run_script")]
    RunScript { base_dir: String, name: String, stdout_is_log: bool, stderr_is_log: bool },
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type")]
pub enum OutboundMsg {
    #[serde(rename = "stdout_normal")]
    StdOutNormal { data: String },

    #[serde(rename = "stdout_log")]
    StdOutLog { data: LogLineInfo },

    #[serde(rename = "stderr_normal")]
    StdErrNormal { data: String },

    #[serde(rename = "stderr_log")]
    StdErrLog { data: LogLineInfo },

    #[serde(rename = "exit")]
    Exit { code: i32 },

    #[serde(rename = "error")]
    Error { msg: String },
}
